pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,418 @@
+defines 4 identifiers, some relative and some absolute:
+
+![A JSON Schema with multiple identifiers](/img/posts/2024/understanding-lexical-dynamic-scopes/scopes-multiple-ids.webp)
+
+[In JSON Schema
+parlance](https://json-schema.org/draft/2020-12/json-schema-core#name-root-schema-and-subschemas-),
+we say that the `$id` keyword introduces a new *schema resource*, and that the
+top level schema resource is referred to as the *root schema resource*.
+
+Consider the following example. This schema consists of 3 schema resources,
+each highlighted using a different color: the root schema resource (red), the
+schema resource at `/properties/foo` (blue), and the schema resource at
+`/properties/bar` (green). Note that the subschema at `/properties/baz` is part
+of the root schema resource, as it does not introduce a new identifier:
+
+![A JSON Schema that consists of 3 schema resources](/img/posts/2024/understanding-lexical-dynamic-scopes/scopes-schema-resources.webp)
+
+Note that children schema resources are not considered part of the parent
+schema resource. For example, in the previous figure, `https://example.com/foo`
+or `https://example.com/bar` are *separate* schema resources and not part of
+the root schema resource, despite their structural relationship.
+
+Schemas as Directed Graphs
+--------------------------
+
+JSON Schema is a recursive data structure. In the context of schema resources,
+this means that a schema resource may introduce nested schema resources (like
+we saw on the previous section) and use referencing keywords (like `$ref`) to
+point to external schema resources, effectively creating a directed graph of
+schema resources. 
+
+Consider the following example. In the top left, a root schema resource named
+`https://example.com/origin` that declares a nested schema resource named
+`https://example.com/nested` (at `/properties/bar`) and references an external
+schema resource named `https://example.com/destination` (from
+`/properties/foo/$ref`). In the bottom left, a root schema resource named
+`https://example.com/destination` that references a nested schema resource
+called `https://example.com/nested-string` (from `/items/$ref`). In the right,
+a directed graph representation of the relationship between these schema
+resources:
+
+![Thinking of a JSON Schema as a directed graph](/img/posts/2024/understanding-lexical-dynamic-scopes/directed-graph.webp)
+
+As you will see, thinking of a schema as a directed graph of schema resources
+greatly helps in understanding both lexical and dynamic scopes.
+
+Lexical Scope
+-------------
+
+Under the graph analogy from the previous section, the lexical scope of a
+schema consists of the node being evaluated. In other words, the lexical scope
+of a schema consists of the entire schema resource of which it is part of. 
+
+Consider the following sequence of examples. In the left, a JSON Schema with a
+single nested schema resource. In the right, the corresponding directed graph
```

Comment:
```suggestion
single nested schema resource. On the right, the corresponding directed graph
```

We typically use "on" when there's only one direction like "top" or "left".

Reply:
@jviotti there are several of these.  I didn't want to mark them all.

Reply:
Ah, got it. Fixing them all now...